src/librustc/middle/infer/mod.rs

```diff
@@ -998,6 +1026,23 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             .collect()
     }
 
+    pub fn type_vars_for_defs(&self,
+                              defs: &[ty::TypeParameterDef<'tcx>])
+                              -> Vec<ty::Ty<'tcx>> {
+        let mut vars = Vec::with_capacity(defs.len());
+
+        for def in defs.iter() {
+            let ty_var = self.next_ty_var();
+            match def.default {
+                None => {},
+                Some(default) => { self.defaults.borrow_mut().insert(ty_var, default); }
```

Comment:
I don't think this is quite right. I think we have to substitute the earlier type variables into the default. In particular, there might be a situation like this:

```rust
impl<T,U=T> Foo<U> for T { .. }
```

My recommendation is to concoct a failing test first that exposes the problem I'm talking about (presuming you agree it exists...)